clang/labs/lab1/prokofev_kirill/RenameIdentificator.cpp

```diff
@@ -0,0 +1,198 @@
+        }
+      }
+    }
+    return true;
+  }
+
+  bool VisitDeclRefExpr(clang::DeclRefExpr *expr) {
+    clang::VarDecl *var = clang::dyn_cast<clang::VarDecl>(expr->getDecl());
+    if (type == IdType::Var && var && var->getName() == oldName) {
+      rewriter.ReplaceText(expr->getSourceRange(), newName);
+    }
+    return true;
+  }
+
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (type == IdType::Class && record->getName() == oldName) {
+      rewriter.ReplaceText(record->getLocation(), newName);
+      const auto *destructor = record->getDestructor();
+      if (destructor) {
+        rewriter.ReplaceText(destructor->getLocation(), oldName.size() + 1,
+                             '~' + newName);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXConstructorDecl(clang::CXXConstructorDecl *constructor) {
+    if (type == IdType::Class) {
+      if (constructor->getNameAsString() == oldName) {
+        rewriter.ReplaceText(constructor->getLocation(), oldName.size(),
+                             newName);
+      }
+    }
+    return true;
+  }
+
+  bool VisitCXXNewExpr(clang::CXXNewExpr *newExpr) {
+    if (type == IdType::Class) {
+      if (newExpr->getConstructExpr()->getType().getAsString() == oldName) {
+        rewriter.ReplaceText(newExpr->getExprLoc(), oldName.size() + 4,
+                             "new " + newName);
+      }
+    }
+    return true;
+  }
+
+  bool save_changes() { return rewriter.overwriteChangedFiles(); }
+
+private:
+  clang::Rewriter rewriter;
+  IdType type;
+  std::string oldName;
+  std::string newName;
+};
+
+class RenameASTConsumer : public clang::ASTConsumer {
+public:
+  explicit RenameASTConsumer(clang::CompilerInstance &CI, IdType type,
+                             clang::StringRef cur_name,
+                             clang::StringRef new_name)
+      : Visitor(clang::Rewriter(CI.getSourceManager(), CI.getLangOpts()), type,
+                cur_name, new_name) {}
+
+  void HandleTranslationUnit(clang::ASTContext &context) override {
+    Visitor.TraverseDecl(context.getTranslationUnitDecl());
+    if (Visitor.save_changes()) {
+      llvm::errs() << "An error occurred while saving changes to a file!\n";
+    }
+  }
+
+private:
+  RenameVisitor Visitor;
+};
+
+class RenamePlugin : public clang::PluginASTAction {
+public:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &CI,
+                    clang::StringRef InFile) override {
+    return std::make_unique<RenameASTConsumer>(CI, type, oldName, newName);
+  }
+
+protected:
+  bool ParseArgs(const clang::CompilerInstance &CI,
+                 const std::vector<std::string> &args) override {
+    std::vector<std::pair<std::string, std::string>> params = {
+        {"type=", ""}, {"oldName=", ""}, {"newName=", ""}};
+
+    if (!args.empty() && args[0] == "help") {
+      PrintHelp(llvm::errs());
+      return false;
+    }
+
+    if (args.size() < 3) {
```

Comment:
What if number of args > 3?

Reply:
An argument error will be displayed.

Reply:
I specify required arguments and `help` argument in the end. So, total number of args is 4, help is present. I expect to see help message only.

Reply:
Fixed the check and added the appropriate test